File: FlowCrypt/src/androidTest/java/com/flowcrypt/email/util/PrivateKeysManager.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util;

import android.content.Context;

import com.flowcrypt.email.api.retrofit.node.gson.NodeGson;
import com.flowcrypt.email.api.retrofit.response.model.node.NodeKeyDetails;
import com.flowcrypt.email.base.BaseTest;
import com.flowcrypt.email.database.dao.KeysDao;
import com.flowcrypt.email.database.dao.source.KeysDaoSource;
import com.flowcrypt.email.database.dao.source.UserIdEmailsKeysDaoSource;
import com.flowcrypt.email.js.UiJsManager;
import com.flowcrypt.email.model.KeyDetails;
import com.flowcrypt.email.security.KeyStoreCryptoManager;
import com.flowcrypt.email.ui.activity.base.BaseActivity;
import com.google.gson.Gson;

import java.io.IOException;
import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.test.internal.runner.junit4.statement.UiThreadStatement;
import androidx.test.platform.app.InstrumentationRegistry;

/**
 * This tool can help manage private keys in the database. For testing purposes only.
 *
 * @author Denis Bondarenko
 * Date: 27.12.2017
 * Time: 17:44
 * E-mail: DenBond7@gmail.com
 */

public class PrivateKeysManager {

  public static void saveKeyFromAssetsToDatabase(String keyPath, String passphrase, KeyDetails.Type type,
                                                 BaseActivity baseActivity) throws Throwable {
    NodeKeyDetails nodeKeyDetails = getNodeKeyDetailsFromAssets(keyPath);
    saveKeyToDatabase(nodeKeyDetails, passphrase, type, baseActivity);
  }

  public static void saveKeyToDatabase(NodeKeyDetails nodeKeyDetails, String passphrase, KeyDetails.Type type,
                                       final BaseActivity baseActivity) throws Throwable {
    final Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();

    KeysDaoSource keysDaoSource = new KeysDaoSource();
    KeyStoreCryptoManager keyStoreCryptoManager = new KeyStoreCryptoManager(InstrumentationRegistry.getInstrumentation()
        .getTargetContext());

    keysDaoSource.addRow(context, KeysDao.generateKeysDao(keyStoreCryptoManager, type, nodeKeyDetails, passphrase));

    new UserIdEmailsKeysDaoSource().addRow(context, nodeKeyDetails.getLongId(),
        nodeKeyDetails.getPrimaryPgpContact().getEmail());

    UiThreadStatement.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        UiJsManager.getInstance(context).getJs().getStorageConnector().refresh(context);
        if (baseActivity != null) {
          baseActivity.restartJsService();
        }
      }
    });
    // Added timeout for a better sync between threads.
    Thread.sleep(3000);
  }

  public static NodeKeyDetails getNodeKeyDetailsFromAssets(String assetsPath) throws IOException {
    Gson gson = NodeGson.getInstance().getGson();
    String json = TestGeneralUtil.readFileFromAssetsAsString(BaseTest.getContext(), assetsPath);
    return gson.fromJson(json, NodeKeyDetails.class);
  }

  @NonNull
  public static ArrayList<NodeKeyDetails> getKeysFromAssets(String[] keysPaths) throws IOException {
    ArrayList<NodeKeyDetails> privateKeys = new ArrayList<>();
    for (String path : keysPaths) {
      privateKeys.add(getNodeKeyDetailsFromAssets(path));
    }
    return privateKeys;
  }

  public static void deleteKey(String keyPath, final BaseActivity baseActivity) throws Throwable {
    NodeKeyDetails nodeKeyDetails = getNodeKeyDetailsFromAssets(keyPath);
    KeysDaoSource keysDaoSource = new KeysDaoSource();

    final Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();

    keysDaoSource.removeKey(context, nodeKeyDetails.getLongId());
    new UserIdEmailsKeysDaoSource().removeKey(context, nodeKeyDetails.getLongId());

    UiThreadStatement.runOnUiThread(new Runnable() {
      @Override
      public void run() {
        UiJsManager.getInstance(context).getJs().getStorageConnector().refresh(context);
        if (baseActivity != null) {
          baseActivity.restartJsService();
        }
      }
    });
    // Added timeout for a better sync between threads.
    Thread.sleep(3000);
  }
}

File: FlowCrypt/src/androidTest/java/com/flowcrypt/email/util/PrivateKeysManager.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util

import androidx.test.platform.app.InstrumentationRegistry
import com.flowcrypt.email.database.FlowCryptRoomDatabase
import com.flowcrypt.email.database.entity.AccountEntity
import com.flowcrypt.email.database.entity.KeyEntity
import com.flowcrypt.email.model.KeyImportDetails
import com.flowcrypt.email.security.KeyStoreCryptoManager
import com.flowcrypt.email.security.model.PgpKeyRingDetails
import com.flowcrypt.email.security.pgp.PgpKey
import org.pgpainless.key.collection.PGPKeyRingCollection

/**
 * This tool can help manage private keys in the database. For testing purposes only.
 *
 * @author Denys Bondarenko
 */
class PrivateKeysManager {
  companion object {
    fun saveKeyFromAssetsToDatabase(
      accountEntity: AccountEntity,
      keyPath: String,
      passphrase: String,
      sourceType: KeyImportDetails.SourceType,
      passphraseType: KeyEntity.PassphraseType = KeyEntity.PassphraseType.DATABASE
    ) {
      val pgpKeyRingDetails = getPgpKeyDetailsFromAssets(keyPath)
      saveKeyToDatabase(accountEntity, pgpKeyRingDetails, passphrase, sourceType, passphraseType)
    }

    fun saveKeyToDatabase(
      accountEntity: AccountEntity,
      pgpKeyRingDetails: PgpKeyRingDetails,
      passphrase: String?,
      sourceType: KeyImportDetails.SourceType = KeyImportDetails.SourceType.EMAIL,
      passphraseType: KeyEntity.PassphraseType = KeyEntity.PassphraseType.DATABASE
    ) {
      val context = InstrumentationRegistry.getInstrumentation().targetContext
      val roomDatabase = FlowCryptRoomDatabase.getDatabase(context)
      val keyEntity = pgpKeyRingDetails
        .copy(passphraseType = passphraseType)
        .toKeyEntity(accountEntity)
        .copy(
          source = sourceType.toString(),
          privateKey = KeyStoreCryptoManager.encrypt(pgpKeyRingDetails.privateKey).toByteArray(),
          storedPassphrase = when (passphraseType) {
            KeyEntity.PassphraseType.DATABASE -> KeyStoreCryptoManager.encrypt(passphrase)
            else -> null
          }
        )
      val existingKey = roomDatabase.keysDao().getKeyByAccountAndFingerprint(
        accountEntity.email,
        pgpKeyRingDetails.fingerprint
      )
      existingKey?.let { roomDatabase.keysDao().delete(it) }
      roomDatabase.keysDao().insert(keyEntity)
      // Added timeout for a better sync between threads.
      Thread.sleep(3000)
    }

    fun savePubKeyToDatabase(assetsPath: String) {
      savePubKeyToDatabase(getPgpKeyDetailsFromAssets(assetsPath))
    }

    fun savePubKeyToDatabase(pgpKeyRingDetails: PgpKeyRingDetails) {
      val context = InstrumentationRegistry.getInstrumentation().targetContext
      val email = requireNotNull(pgpKeyRingDetails.getPrimaryInternetAddress()).address.lowercase()
      val roomDatabase = FlowCryptRoomDatabase.getDatabase(context)
      if (roomDatabase.recipientDao().getRecipientByEmail(email) == null) {
        roomDatabase.recipientDao().insert(requireNotNull(pgpKeyRingDetails.toRecipientEntity()))
      }
      roomDatabase.pubKeyDao().insert(pgpKeyRingDetails.toPublicKeyEntity(email))
      // Added timeout for a better sync between threads.
      Thread.sleep(500)
    }

    fun getPgpKeyDetailsFromAssets(
      assetsPath: String,
      onlyPrivate: Boolean = false
    ): PgpKeyRingDetails {
      return getPgpKeyDetailsListFromAssets(assetsPath, onlyPrivate).first()
    }

    fun getPgpKeyDetailsListFromAssets(
      assetsPath: String,
      onlyPrivate: Boolean = false
    ): List<PgpKeyRingDetails> {
      val parsedCollections =
        PgpKey.parseKeys(TestGeneralUtil.readFileFromAssetsAsStream(assetsPath))

      if (onlyPrivate) {
        val onlyPrivateKeysCollection = PgpKey.ParseKeyResult(
          PGPKeyRingCollection(
            parsedCollections.pgpKeyRingCollection
              .pgpSecretKeyRingCollection.keyRings.asSequence().toList(), false
          )
        )

        return onlyPrivateKeysCollection.pgpKeyDetailsList
      } else {
        return parsedCollections.pgpKeyDetailsList
      }
    }

    fun getKeysFromAssets(
      keysPaths: Array<String>,
      onlyPrivate: Boolean = false
    ): ArrayList<PgpKeyRingDetails> {
      val privateKeys = ArrayList<PgpKeyRingDetails>()
      keysPaths.forEach { path ->
        privateKeys.addAll(getPgpKeyDetailsListFromAssets(path, onlyPrivate))
      }
      return privateKeys
    }

    fun deleteKey(accountEntity: AccountEntity, keyPath: String) {
      val pgpKeyRingDetails = getPgpKeyDetailsFromAssets(keyPath)
      deleteKey(accountEntity, pgpKeyRingDetails)
    }

    fun deleteKey(accountEntity: AccountEntity, pgpKeyRingDetails: PgpKeyRingDetails) {
      val context = InstrumentationRegistry.getInstrumentation().targetContext
      val roomDatabase = FlowCryptRoomDatabase.getDatabase(context)
      pgpKeyRingDetails.fingerprint.let {
        roomDatabase.keysDao().deleteByAccountAndFingerprint(accountEntity.email, it)
      }

      // Added timeout for a better sync between threads.
      Thread.sleep(3000)
    }
  }
}
